### ISSUE TITLE
editor with pango text display glitches

### ISSUE THREAD

**Issue by Albrecht-S** [fltk/fltk#59]
Editing the text file [pango_editor.txt](https://github.com/fltk/fltk/files/4268696/pango_editor.txt) with FLTK's `test/editor` has two issues:

1. The URL in line 1 (displayed as comment) is truncated (should be '.html')
2. `\n` in line 5 is displayed as `¥n` (Yen + 'n').

(1) ISTR that we had issues with text in italics, but I can't remember if or how it was resolved. Maybe it's still an open STR...

(2) See attached screenshot. I'm not sure if this is a `pango` issue though: opening the same file in the system editor `xed` seems to have the same issue. Inserting another character like `'x'` before `'\n'` fixes the issue (a space does not fix it). Can anybody confirm this?

As noted in the text file, the text in question was copied from the web site mentioned in the [URL in the text file](http://python-future.org/compatible_idioms.html#unicode-text-string-literals).

Screenshot (web site in the background, text editor below):
![pango_editor_screenshot](https://user-images.githubusercontent.com/26235500/75581279-79d09480-5a69-11ea-8864-96079199406c.png)

**Comment by ManoloFLTK** [comment]
Issue 1) is related to clipping when Fl_Text_Display draws and
occurs identically on all platforms (macOS screenshot below).

Issue 2) does not occur with macOS.

<img width="418" alt="image" src="https://user-images.githubusercontent.com/41016272/75602940-38bc9c80-5aca-11ea-89d1-e731a17457a5.png">

**Comment by ManoloFLTK** [comment]
I don't see here (tested on Ubuntu in VirtualBox and on XQuartz == X11-on-macOS)
issue #2 (see attached Ubuntu screenshot)  :

<img width="663" alt="image" src="https://user-images.githubusercontent.com/41016272/75603176-702c4880-5acc-11ea-868e-756c171fedf3.png">

**Comment by ManoloFLTK** [comment]
With file pango_editor.txt  :

<img width="666" alt="image" src="https://user-images.githubusercontent.com/41016272/75603303-55a69f00-5acd-11ea-9348-ec6356785ee7.png">

**Comment by ManoloFLTK** [comment]
and with file misc/lorem_ipsum.txt from the FLTK source tree:

<img width="706" alt="image" src="https://user-images.githubusercontent.com/41016272/75603434-7ae7dd00-5ace-11ea-90ed-2435f472205d.png">

**Comment by ManoloFLTK** [comment]
I also tried with an old Debian version
 lsb_release -a
Distributor ID:	Debian
Description:	Debian GNU/Linux 8.11 (jessie)
Release:	8.11
Codename:	jessie
and FLTK 1.4 also draws correctly file pango_editor.txt.

I'm very puzzled by your finding with Linux Mint.

**Comment by Albrecht-S** [comment]
FWIW, here is some info about my current libpango* versions:
```
$ ldd bin/examples/editor | grep 'pango\|xft'
	libpango-1.0.so.0 => /usr/lib/x86_64-linux-gnu/libpango-1.0.so.0 (0x00007fee66f78000)
	libpangoxft-1.0.so.0 => /usr/lib/x86_64-linux-gnu/libpangoxft-1.0.so.0 (0x00007fee66d6f000)
	libpangoft2-1.0.so.0 => /usr/lib/x86_64-linux-gnu/libpangoft2-1.0.so.0 (0x00007fee65278000)
```
... and my installed libpango* packages (list edited/reduced):
```
$ apt list --installed | grep libpango
libpango-1.0-0/bionic-updates,bionic-security,now 1.40.14-1ubuntu0.1 amd64 [installed]
libpango1.0-dev/bionic-updates,bionic-security,now 1.40.14-1ubuntu0.1 amd64 [installed]
libpangoft2-1.0-0/bionic-updates,bionic-security,now 1.40.14-1ubuntu0.1 amd64 [installed]
libpangoxft-1.0-0/bionic-updates,bionic-security,now 1.40.14-1ubuntu0.1 amd64 [installed]
```
Maybe you can compare with your versions. (?)
I'll do some additional tests on another system when time permits...

**Comment by ManoloFLTK** [comment]
On a Debian buster system running Linux 4.9 I get here:
ldd editor | grep 'pango\|xft'
	libpango-1.0.so.0 => /usr/lib/x86_64-linux-gnu/libpango-1.0.so.0 (0x00007f8656731000)
	libpangoxft-1.0.so.0 => /usr/lib/x86_64-linux-gnu/libpangoxft-1.0.so.0 (0x00007f8656726000)
	libpangoft2-1.0.so.0 => /usr/lib/x86_64-linux-gnu/libpangoft2-1.0.so.0 (0x00007f8655cc2000)

and

apt list --installed | grep libpango
libpango-1.0-0/testing,now 1.42.4-6 amd64  [installé, automatique]
libpango1.0-0/testing,now 1.42.4-6 amd64  [installé]
libpango1.0-dev/testing,now 1.42.4-6 amd64  [installé]
libpangocairo-1.0-0/testing,now 1.42.4-6 amd64  [installé, automatique]
libpangoft2-1.0-0/testing,now 1.42.4-6 amd64  [installé, automatique]
libpangomm-1.4-1v5/stable,now 2.40.1-3 amd64 [installé, pouvant être mis à jour vers : 2.42.0-2]
libpangox-1.0-0/testing,unstable,stable,now 0.0.2-5+b2 amd64  [installé]
libpangoxft-1.0-0/testing,now 1.42.4-6 amd64  [installé]

**Comment by ManoloFLTK** [comment]
On my old Debian jessie that also draws file pango_editor.txt correctly
pango libs are older:

 apt list --installed | grep libpango
libpango-1.0-0/oldoldstable,now 1.36.8-3 i386 [installed]
libpango1.0-0/oldoldstable,now 1.36.8-3 i386 [installed,automatic]
libpango1.0-dev/oldoldstable,now 1.36.8-3 i386 [installed,automatic]
libpangocairo-1.0-0/oldoldstable,now 1.36.8-3 i386 [installed]
libpangoft2-1.0-0/oldoldstable,now 1.36.8-3 i386 [installed]
libpangomm-1.4-1/oldoldstable,now 2.34.0-1.1 i386 [installed,automatic]
libpangox-1.0-0/oldoldstable,now 0.0.2-5 i386 [installed]
libpangoxft-1.0-0/oldoldstable,now 1.36.8-3 i386 [installed]

**Comment by ManoloFLTK** [comment]
The glitch you report is strange because I expect both editor of FLTK and, say, gedit
of gnome to call the same pango libraries to draw the same text. Precisely, FLTK calls
functions pango_layout_set_text() followed by pango_xft_render_layout() do draw all text.

**Comment by Albrecht-S** [comment]
I wrote in my original post:

> I'm not sure if this is a pango issue though: opening the same file in the system editor xed seems to have the same issue.

Now I tested different editors with the following results:

- fltk: fault
- xed: fault
- gedit: fault
- scite: OK
- LibreOffice Writer: OK

where `'fault'` means the **same** glitch as reported and `'OK'` means correct display.

After all I assume that this is really a pango issue of my specific version and we should end discussion of this part (2) because it's not a FLTK issue.

Regarding part 1 (text truncation): I found that this is covered by [STR 3364](https://www.fltk.org/str.php?L3364).

@ManoloFLTK Thanks for your comments and testing.

I'm closing this issue now.
